[SourceKit] ⚡️Fast code completion within function bodies

rintaro opened the pull request:
Reuse `ASTContext`(`CompilerInstance`) from the previous completions when possible.

`ide::CompletionInstance` manages `CompilerInstance` for completion like requests (i.e. code completion, type context info, and conforming method list).
It vends the cached `CompilerInstance` if:
- This functionality is enabled, and
- The compiler arguments (i.e. `CompilerInvocation`) has not changed, and
- The primary file is the same, and
- The completion happens inside function bodies in both previous and current completion, and
- The interface hash of of the primary file has not changed

Since cached instance already have `AST` including imported modules, we just run the second pass.

Otherwise, `CompletionInstance` vends new `CompilerInstance`, so we run both first and second pass using that instance. This instance is cached for the next completion request.

NOTE: currently it's disabled by default.

For example, using this simple example:
```
func test(x: Int) {
  x.#^COMPLETE^#
}
```
```
# Disabled
$ sourcekitd-test -req=complete -pos=2:5 $PWD/test.swift -repeat-request=10 -dont-print-request -dont-print-response -time-request -- $PWD/test.swift
request time: 109.469 ms
request time: 102.269 ms
request time: 101.260 ms
request time: 99.622 ms
request time: 100.751 ms
request time: 100.154 ms
request time: 99.670 ms
request time: 100.965 ms
request time: 101.426 ms
request time: 100.784 ms
```
```
# Enabled
$ sourcekitd-test -req=complete -req-opts=reuseastcontext=1 -pos=2:5 $PWD/test.swift -repeat-request=10 -dont-print-request -dont-print-response -time-request -- $PWD/test.swift
request time: 108.051 ms
request time: 1.716 ms
request time: 1.553 ms
request time: 1.559 ms
request time: 1.466 ms
request time: 1.497 ms
request time: 1.445 ms
request time: 1.544 ms
request time: 1.447 ms
request time: 1.506 ms
```

rdar://problem/20787086

rintaro committed:
[SourceKit] Support VFS in 'typecontextinfo' and 'conformingmethods' req

Align with code completion.

rintaro committed:
[CodeCompletion] Align swiftCodeCompleteImpl() with other similar funcs

Specifically, align with swiftTypeContextInfoImpl() and
swiftConformingMethodListImpl()

rintaro committed:
[SourceKit] Reuse compiler instance between multiple completion

- Introduce ide::CompletionInstance to manage CompilerInstance
- `CompletionInstance` vends the cached CompilerInstance when:
-- The compiler arguments (i.e. CompilerInvocation) has has not changed
-- The primary file is the same
-- The completion happens inside function bodies in both previous and
   current completion
-- The interface hash of the primary file has not changed
- Otherwise, it vends a fresh CompilerInstance and cache it for the next
  completion

rdar://problem/20787086

rintaro committed:
[CodeCompletion] Use offsets in the buffer for second pass state

So that we can use the different the buffer for the second pass from the
first pass.

rintaro committed:
[SourceKit] Add option to enable ASTContext reusing for code completion

rintaro committed:
[SourceKit] Calculate CompilerInvocation hash

rintaro committed:
[SourceKit] Add a test case for fast code completion

rintaro committed:
[SourceKit] Limit compiler instance reuse count for completions

rintaro commented:
@swift-ci Please test

rintaro commented:
@swift-ci Please test Linux

rintaro committed:
[SourceKit] Add test file for fast completion

rintaro commented:
Forgot to add a test file.
@swift-ci Please test

benlangmuir commented on include/swift/IDE/CompletionInstance.h:
```diff
@@ -0,0 +1,69 @@
+  unsigned CurrentASTReuseCount = 0;
+
+  swift::CompilerInstance *
+  getReusingCompilerInstance(const swift::CompilerInvocation &Invocation,
```
How about `getCachedCompilerInstance`  or `getReusableCompilerInstance`?  This could use a doc comment.

benlangmuir commented on include/swift/IDE/CompletionInstance.h:
```diff
@@ -0,0 +1,69 @@
+                               unsigned &Offset,
+                               llvm::StringRef bufferIdentifier);
+
+class CompletionInstance {
```
This could use a doc comment to explain its purpose.

benlangmuir commented on include/swift/Parse/LocalContext.h:
```diff
@@ -19,6 +19,7 @@
 #define SWIFT_PARSE_LOCALCONTEXT_H
 
 #include "llvm/ADT/DenseMap.h"
+#include "swift/AST/Identifier.h"
```
What is this change for?

rintaro replied:
This file is not included by "CompletionInstance.cpp" via "PersistentParserState.h" too. Since this file uses `Identifier`, it causes a compiler error without this.

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -0,0 +1,360 @@
+/// Returns index number of \p D in \p Decls . If it's not found, returns ~0.
+template <typename Range>
+unsigned findIndexInRange(Decl *D, const Range &Decls) {
+  unsigned N = 0;
```
Is `Range` always random-access? If so, maybe simplify to

```c++
auto I = std::find(Decls.begin(), Decls.end(), D);
return I == Decls.end() ? ~0u : std::distance(Decls.begin(), I);
```

rintaro replied:
`IterableDeclContext::getMembers()` returns `DeclRange` (`iterator_range<DeclIterator>`) which is unfortunately non-random-access.

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -0,0 +1,360 @@
+/// This assumes the AST which contains \p DC has exact the same structure with
+/// \p SF.
+/// FIXME: This doesn't support IfConfigDecl blocks. If \p DC is in an inactive
+///        config block, this function probably returns false.
```
"probably"?

rintaro replied:
Probably "always" :)

(will fix)

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -0,0 +1,360 @@
+
+/// Return the element at \p N in \p Decls .
+template <typename Range> Decl *getElementAt(const Range &Decls, unsigned N) {
+  auto I = Decls.begin();
```
`assert(std::distance(Decls.begin(), Decls.end()) > N)` ?

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -0,0 +1,360 @@
+/// This should take all options affecting completion results into account.
+/// If the hashes between multiple completion invocation are different, we
+/// cannot reuse the CompilerInstance.
+static llvm::hash_code calculateInvocationHash(const CompilerInvocation &Inv) {
```
How does this compare to `getPCHHash()`?  Can we reuse some of that logic?

rintaro replied:
`getPCHHash()` doesn't take `FrontendOptions` into account. That means it's not updated even when other files are added to the project.

> Can we reuse some of that logic?

Probably yes. But I haven't looked into it

rintaro replied:
This function is now completely gone.

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -0,0 +1,360 @@
+    return nullptr;
+
+  if (calculateInvocationHash(Invocation) !=
+      calculateInvocationHash(CachedCI->getInvocation()))
```
Why do we support non-identical compiler invocations? When does that happen in practice that we care about?

rintaro replied:
> Why do we support non-identical compiler invocations?

What do you mean? Are you suggesting to implement `CompilerInvocation::operator==` unless support non-identical compiler invocations?

benlangmuir replied:
If we only care about equality, could we compare the compiler arguments directly as strings instead? What I'm concerned about is enumerating the options in this file is hard to keep up to date when we add new options.

rintaro replied:
I see. I completely agree with your concern.
Just comparing `Args` was what I implemented at first. The reason I stop doing that was I just didn't want to pass them to this place. But I will reconsider that.

rintaro replied:
Modified to remember the hash of the arguments of the invocation.

benlangmuir commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3776,7 +3777,7 @@ std::vector<Decl *> Parser::parseDeclListDelayed(IterableDeclContext *IDC) {
   }
 
-  auto BeginParserPosition = getParserPosition({BodyRange.Start,BodyRange.End});
+  auto BeginParserPosition = getParserPosition({BodyRange.Start, SourceLoc()});
```
What is this change about?

rintaro replied:
`getParserPosition()` receives `PersistentParserState::ParserPos` which receives the token location and its *previous* location. That loc is to be restored to `PreviousLoc` in parser. So `BodyLoc.End` was just a mistake.

I admit that this change is not needed to this PR though.

rintaro commented on tools/SourceKit/lib/SwiftLang/SwiftLangSupport.h:
```diff
@@ -294,6 +295,7 @@ class SwiftLangSupport : public LangSupport {
   ThreadSafeRefCntPtr<SwiftCustomCompletions> CustomCompletions;
   std::shared_ptr<SwiftStatistics> Stats;
   llvm::StringMap<std::unique_ptr<FileSystemProvider>> FileSystemProviders;
+  std::unique_ptr<swift::ide::CompletionInstance> CompletionInst;
```
Ah, I see. I will change this to hold and vend `shared_ptr`.

rintaro committed:
[CodeCompletion] Add doc-comment for CompletionInstance

rintaro committed:
[CodeCompletion] Rename getReusingCompilerInstance()

to getCachedCompilerInstance()

rintaro committed:
[SourceKit] Hold CompletionInstance with std::shared_ptr

So we can pin it on the threads.

rintaro committed:
[CodeCompletion] Use the arguments to check the equality of the invocation

Checking "selected" properties of CompilerInvocation is hard to keep it
up to date. If the arguments are the same, the invocations are the same.

rintaro commented:
@swift-ci Please test

rintaro commented:
@swift-ci Please test Linux

benlangmuir reviewed (commented):
It would be good to have some additional tests:
* Compiler argument changed
* Source changed in a compatible way -> reuse
* Source changed in an incompatible way -> new AST

You can modify the text that is passed without changing the compiler arguments by using `-text-input` option to `sourcekitd-test` I think.

rintaro committed:
[SourceKit] Add more test cases for fast completions

rintaro commented on tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp:
```diff
@@ -970,6 +972,8 @@ static int handleTestInvocation(TestOptions Opts, TestOptions &InitOpts) {
     sourcekitd_request_dictionary_set_string(Req, KeySourceText,
                                              Opts.SourceText->c_str());
+    sourcekitd_request_dictionary_set_string(Req, KeySourceFile,
+                                             SemaName.c_str());
```
This change was needed to use `-name` as the filename (`key.sourcefile`).

rintaro commented:
@swift-ci Please smoke test

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftCompletion.cpp:
```diff
@@ -129,104 +130,102 @@ static bool swiftCodeCompleteImpl(
-  Invocation.setCodeCompletionPoint(NewBuffer.get(), CodeCompletionOffset);
+  // Pin completion instance.
+  auto CompletionInst =  Lang.getCompletionInstance();
+  CompilerInstance *CI = CompletionInst->getCompilerInstance(
```
Oh hmm, this probably isn't thread safe.  I don't think anything prevents you from making multiple code-completion requests (or other completion-like requests) at the same time.  In sourcekitd-test, for example if you add `-async` and then make a bunch of requests in a row they probably overlap.

Perhaps we should use `shared_ptr<CompilerInstance>` so that if we renew it can't affect an ongoing completion?

rintaro replied:
Thanks! I confirmed the crashes with `-async`. Will fix.

rintaro committed:
[SourceKit/CodeCompletion] Use callback function to run the second pass

To controls the lifetime of CompilerInstance within CompletionIntance.

- Prevent from using the same CompilerInstance from multiple completion
  requests
- Separate the stacktrace between "fast" and "normal" completions
- Further code consolidation between various completion-like requests

rintaro commented:
@benlangmuir The last commit changes the structure of the process. So `CompletionIntance` take the control of the lifetime of `CompilerInstace`.
`CompletionIntance::performOperation()` receives a callback `void(CompilerInstance &)`, and call it with the prepared compiler instance. This change has several benefits:

- Prevent the same instance is used from multiple threads
- Stacktrace is distinguishable between "fast" and "normal" mode
- Consolidate more code between completion / type context / conforming method list

What do you think?

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[CodeCompletion] Pack cached instance information into single struct

To prevent them from out-of-sync.

rintaro commented:
Modified the way of storing the cached information. It should prevent `CompilerInstance`/`ArgHash`/`ReuseCount` from being out-of-sync.
@swift-ci Please smoke test

benlangmuir commented:
Ah good point, a shared_ptr is wrong since you actually can't share this safely across threads. Note: your current code is still technically racy because the `std::swap()` of the state itself is not atomic.

---

So if you have racing completions, one will win and get to reuse the AST, and the next one will go through the slow path.  Is that better than blocking on the first one?  If you're reusing ASTs it seems like it would usually be better to wait since you have a chance to hit the fast path on every completion.  Assuming the reuse case is 10x faster and that the number of concurrent completions is usually very small, you would be better off waiting.

If you're *not* reusing ASTs, then I guess running the completions in parallel might be better in some instances - or at least it matches the current behaviour.

rintaro committed:
[CodeCompletionn] Block completions in other threads

if fast-completion is enabled. So they have higher chance to use the cached
completion instance.
If it's disabled, don't block, use an ephemeral instance so we can peform
multiple completions simultaneously.

rintaro commented:
Fair enough. I update the PR to block other threads using `std::mutex`.
@swift-ci Please smoke test

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -278,8 +274,9 @@ bool CompletionInstance::performNewOperation(
-    CachedInst.swap(TheInstance);
+    CachedCI = std::move(TheInstance);
+    CachedArgHash = ArgsHash;
+    CachedReuseCound = 0;
```
Typo: "Cound" vs "Count"

benlangmuir commented on include/swift/IDE/CompletionInstance.h:
```diff
@@ -0,0 +1,97 @@
+      llvm::function_ref<void(CompilerInstance &)> Callback);
+
+public:
+  void setEnableASTCaching(bool Flag) { EnableASTCaching = Flag; }
```
Even if this were done atomically it would not be correct for racing completions.  Can we pass `EnableASTCaching` in as an argument to `performOperation` so that is is associated with a specific completion?

rintaro replied:
Done! How does it look now?

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -0,0 +1,324 @@
+
+  // If AST caching is enabled, block completions in other threads. So they
+  // have higher chance to use the cached completion instance.
+  Optional<std::lock_guard<std::mutex>> lock;
```
This is fun, but maybe too magical?  Could we instead separate these code-paths, like

```c++
if (EnableASTCaching) {
  // Compute the signature of the invocation.
  llvm::hash_code ArgsHash(0);
  for (auto arg : Args)
     ArgsHash = llvm::hash_combine(ArgsHash, StringRef(arg));

  // Concurrent completions will block so that they have a higher chance to use the cached completion instance.
  std::lock_guard<std::mutex> lock(mtx);
  if (performCachedOperaitonIfPossible(...))
    return true;
  if (performNewOperation(...))
     return true;

} else {
  // Concurrent completions may happen in parallel when caching is disabled.
  if (performNewOperation(Invocation, ArgsHash, FileSystem, completionBuffer,
                           Offset, Error, DiagC, Callback))
     return true;
}
```

rintaro committed:
[SourceKit] Pass 'EnableASTCaching' flag as an argument

so that it is associated with a specific completion.

rintaro commented:
@swift-ci Please smoke test
